packages/rabbitmq/src/amqp/connection.ts

```diff
@@ -23,43 +25,133 @@ const defaultConfig = {
-  private _connection!: amqplib.Connection;
-  private _channel!: amqplib.Channel;
-  private config: Required<RabbitMQConfig>;
+  private readonly messageSubject = new Subject<CorrelationMessage>();
+  private readonly config: Required<RabbitMQConfig>;
+  private readonly logger: Logger;
+  private _connection!: amqpcon.AmqpConnectionManager;
+  private _channel!: amqpcon.ChannelWrapper;
+  private _rawChannel?: amqplib.Channel;
+
   constructor(config: RabbitMQConfig) {
     this.config = { ...defaultConfig, ...config };
+    this.logger = new Logger(AmqpConnection.name);
   }
 
   get channel() {
     return this._channel;
   }
 
+  get rawChannel() {
+    return this._rawChannel;
+  }
+
   get connection() {
     return this._connection;
   }
 
-  public async init() {
-    this._connection = await amqplib.connect(this.config.uri);
-    this._channel = await this._connection.createChannel();
-
-    await Promise.all(
-      this.config.exchanges.map(async x =>
-        this._channel.assertExchange(
-          x.name,
-          x.type || this.config.defaultExchangeType,
-          x.options
-        )
+  public async init(): Promise<void> {
+    const {
+      wait = false,
+      timeout = 3000,
+      reject = false
+    } = this.config.connectionInit;
+
+    if (!wait) {
+      return this.initCore();
+    }
+
+    const connectAsync = new Promise(res => this.initCore(() => res()));
+
+    const timeoutAsync = new Promise((res, rej) =>
+      setTimeout(
+        () => (reject ? rej('Failed to connect to a RabbitMQ broker') : res()),
+        timeout
       )
     );
 
-    await this.channel.prefetch(this.config.prefetchCount);
+    return Promise.race<any>([connectAsync, timeoutAsync]);
+  }
+
+  private async initCore(firstInitCallback?: () => void): Promise<void> {
+    this._connection = amqpcon.connect(
+      Array.isArray(this.config.uri) ? this.config.uri : [this.config.uri],
+      this.config.connectionManager
+    );
+
+    this.logger.log('Trying to connect to a RabbitMQ broker');
+    this._connection.on('connect', () => {
+      this.logger.log('Successfully connected to a RabbitMQ broker');
+    });
+
+    this._channel = this._connection.createChannel({
+      name: AmqpConnection.name
+    });
+
+    if (firstInitCallback) {
+      this._channel.once('connect', () => firstInitCallback());
+    }
+    this._channel.on('connect', () =>
+      this.logger.log('Successfully connected a RabbitMQ channel')
+    );
+    this._channel.on('error', (err, { name }) => {
+      this._rawChannel = undefined;
+      this.logger.log(
+        `Failed to setup a RabbitMQ channel - name: ${name} / error: ${
+          err.message
+        } ${err.stack}`
+      );
+    });
+    this._channel.on('close', () => {
+      this._rawChannel = undefined;
+      this.logger.log('Successfully closed a RabbitMQ channel');
+    });
+
+    await this._channel.addSetup(this.setupInitChannel.bind(this));
```

Comment:
I'd rather have an arrow function here instead of doing `.bind(this)` it's a bit of personal preference but I don't like having to spend cycles worrying about `this` context with JS

Reply:
Done